src/backend/resources.rs

```diff
@@ -0,0 +1,270 @@
+//! Resource estimation backend for the Calyx compiler.
+//! Transforms an [`ir::Context`](crate::ir::Context) into a CSV that
+//! counts the number of different primitives in a program and loosely
+//! estimates the total size of the generated hardware.
+
+use std::io::Write;
+use std::vec;
+use std::{collections::HashMap, io};
+
+use crate::backend::traits::Backend;
+use calyx::ir::Component;
+use calyx::{
+    errors::CalyxResult,
+    ir::{self, Binding, CellType, Id},
+    utils::OutputFile,
+};
+
+#[derive(Default)]
+pub struct ResourcesBackend;
+
+impl Backend for ResourcesBackend {
+    fn name(&self) -> &'static str {
+        "resources"
+    }
+
+    /// OK to run this analysis on any Calyx program
+    fn validate(_ctx: &ir::Context) -> CalyxResult<()> {
+        Ok(())
+    }
+
+    /// Don't need to take care of this for this pass
+    fn link_externs(
+        _ctx: &ir::Context,
+        _file: &mut OutputFile,
+    ) -> CalyxResult<()> {
+        Ok(())
+    }
+
+    fn emit(ctx: &ir::Context, _file: &mut OutputFile) -> CalyxResult<()> {
+        let main_comp = ctx
+            .components
+            .iter()
+            .find(|comp| comp.name == ctx.entrypoint)
+            .unwrap();
+
+        let count_map: HashMap<(Id, Binding), u32> = gen_count_map(main_comp);
+
+        write_csv(count_map.clone());
+
+        estimated_size(count_map);
+
+        Ok(())
+    }
+}
+
+/// Counts the number of each primitive with a given set of parameters
+/// in the program with entrypoint `main_comp`.
+/// TODO: handle multi-component designs
+fn gen_count_map(main_comp: &Component) -> HashMap<(Id, Binding), u32> {
+    let mut count_map: HashMap<(Id, Binding), u32> = HashMap::new();
+    for cell in main_comp.cells.iter() {
+        let cell_ref = cell.borrow();
+        match &cell_ref.prototype {
+            CellType::Primitive {
+                name,
+                param_binding,
+                ..
+            } => match name {
+                _ if name == "std_reg" => {
```

Comment:
Instead of a long `if` chain, how about creating a `vec![]` of all the names for which you want to calculate counts and doing something like:
```
if cells.contains(name) {
  count_map.entry(...).or_default() += 1
}
```

Reply:
having a constant hashset would also work though admittedly the lookup speed is unlikely to be relevant in this case.

To explain in a bit more detail, this inner match is unnecessary since all the cases are irrefutable (`_` matches everything) the only thing distinguishing each case is the additional guard. Given that, the inner match could  be entirely replaced with a chain of connected if-elses. This is of course equivalent to what was written but is easier to read as it doesn't need the to include the irrefutable pattern match.

Now as Rachit pointed out the content of each arm body is essenitally equivalent with the only exception being which string was matched, so it suffices to instead have a collection (hashset, vec, choose your poison) and to check whether the name is included and increment the entry count via the name variable rather than the hardcoded string which matched. Effectivley reducing this into match/no-match rather than separately needing to consider each primitive.